src/Tmds.ExecFunction/ExecFunction.cs

```diff
@@ -282,12 +282,109 @@ private static string[] GetApplicationArguments()
                 string commandLine = Marshal.PtrToStringAuto(ptr);
                 return CommandLineToArgs(commandLine);
             }
+            else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+            {
+                // The following logic is based on https://gist.github.com/nonowarn/770696
+                // Set up the mib array and the query for process maximum args size
+                var mib = new int[3];
+                int mibLength = 2;
+                mib[0] = MACOS_CTL_KERN;
+                mib[1] = MACOS_KERN_ARGMAX;
+
+                int size = IntPtr.Size / 2;
+                int argmax = 0;
+                var argv = new List<string>();
+
+                var mibHandle = GCHandle.Alloc(mib, GCHandleType.Pinned);
+                try
+                {
+                    var mibPtr = mibHandle.AddrOfPinnedObject();
+                    
+                    // Get the process args size
+                    SysCtl(mibPtr, mibLength, ref argmax, ref size, IntPtr.Zero, 0);
+
+                    // Get the PID so we can query this process' args
+                    var pid = Process.GetCurrentProcess().Id;
+
+                    // Now read the process args into the allocated space
+                    IntPtr procargs = Marshal.AllocHGlobal(argmax);
+                    try
+                    {
+                        mib[0] = MACOS_CTL_KERN;
+                        mib[1] = MACOS_KERN_PROCARGS2;
+                        mib[2] = pid;
+                        mibLength = 3;
+
+                        SysCtl(mibPtr, mibLength, procargs, ref argmax, IntPtr.Zero, 0);
+
+                        // The memory block we're reading is a series of null-terminated strings
+                        // that looks something like this:
+                        //
+                        // | argc      | <int>
+                        // | exec_path | ... \0
+                        // | argv[0]   | ... \0
+                        // | argv[1]   | ... \0
+                        //   ...
+
+                        // Read argc
+                        var argc = Marshal.ReadInt32(procargs);
+
+                        // Skip over argc
+                        var argvPtr = IntPtr.Add(procargs, sizeof(int));
+
+                        // Skip over exec_path
+                        var offset = 0;
+                        while (Marshal.ReadByte(argvPtr, offset) != 0) { offset++; }
+                        while (Marshal.ReadByte(argvPtr, offset) == 0) { offset++; }
```

Comment:
I notice https://gist.github.com/nonowarn/770696 has similar code.
It's weird that we're potentially reading multiple `\0` here.
I think this may cause us to unintentionally move past empty `argv` values, and then even start reading memory past the `argv` array. Does that seem like a valid concern?
Does this work when reading a single `\0` also?

Reply:
Sr my bad I should have commented this a bit better. The reason for this is that `exec_path` is padded with x amount of `\0`. For me, that padding is always exactly 4bytes long. I haven't really found much information about to the length and possible values of this padding. I know that most related implementations also just keep reading until the first none `\0` byte after which they start reading argv. I don't think the first argv can actually contain an empty value, although I'm not truly convinced this statement is true though😅 . 

The layout is actually more like this:

```
| argc      | <int> is always 4 bytes long even on 64bit architectures
| exec_path | ... \0\0\0\0 * ?
| argv[0]   | ... \0
| argv[1]   | ... \0
| argv[2]   | ... \0
...
|env[0]     | .... \0  (VALUE=SOMETHING\0)
...
```
Did a couple of expirements and this is the layout with different arguments:

```sh
nickvd@mac-and-cheese: /tmp/macos-mem
λ /Users/nickvd/.dotnet/dotnet exec /private/tmp/macos-mem/bin/Debug/netcoreapp3.1/macos-mem.dll --help --version --hello=world --value 1 /test 2 -t hello world
-----------------------
| argc      | 13
| exec_path | /Users/nickvd/.dotnet/dotnet \0\0\0\0
| argv[0]   | /Users/nickvd/.dotnet/dotnet \0
| argv[1]   | exec \0
| argv[2]   | /private/tmp/macos-mem/bin/Debug/netcoreapp3.1/macos-mem.dll \0
| argv[3]   | --help \0
| argv[4]   | --version \0
| argv[5]   | --hello=world \0
| argv[6]   | --value \0
| argv[7]   | 1 \0
| argv[8]   | /test \0
| argv[9]   | 2 \0
| argv[10]   | -t \0
| argv[11]   | hello \0
| argv[12]   | world \0

nickvd@mac-and-cheese: /tmp/macos-mem
λ dotnet exec /private/tmp/macos-mem/bin/Debug/netcoreapp3.1/macos-mem.dll --help --version --hello=world --value 1 /test 2 -t hello world
-----------------------
| argc      | 13
| exec_path | /Users/nickvd/.dotnet/dotnet \0\0\0\0
| argv[0]   | dotnet \0
| argv[1]   | exec \0
| argv[2]   | /private/tmp/macos-mem/bin/Debug/netcoreapp3.1/macos-mem.dll \0
| argv[3]   | --help \0
| argv[4]   | --version \0
| argv[5]   | --hello=world \0
| argv[6]   | --value \0
| argv[7]   | 1 \0
| argv[8]   | /test \0
| argv[9]   | 2 \0
| argv[10]   | -t \0
| argv[11]   | hello \0
| argv[12]   | world \0
```

Reply:
> I don't think the first argv can actually contain an empty value, although I'm not truly convinced this statement is true thoughsweat_smile .

Yes, the first argument is the program name. I'm ok assuming it can't be empty.